File: tests/Aiko.Core.Tests/IslandPlacementTests.cs
using Aiko.Core;

namespace Aiko.Core.Tests;

public class IslandPlacementTests
{
    /// A 1920 by 1080 screen with the taskbar at the bottom.
    private static readonly Box Work = new(0, 0, 1920, 1040);

    private const double Width = 160;
    private const double Height = 34;

    [Fact]
    public void An_island_near_the_top_sticks_to_the_top()
    {
        var dropped = new Box(880, 12, Width, Height);

        Assert.Equal(ScreenEdge.Top, IslandPlacement.Nearest(dropped, Work).Edge);
    }

    [Fact]
    public void An_island_near_the_right_sticks_to_the_right()
    {
        var dropped = new Box(1740, 500, Width, Height);

        Assert.Equal(ScreenEdge.Right, IslandPlacement.Nearest(dropped, Work).Edge);
    }

    [Fact]
    public void An_island_near_the_bottom_sticks_to_the_bottom()
    {
        var dropped = new Box(400, 1000, Width, Height);

        Assert.Equal(ScreenEdge.Bottom, IslandPlacement.Nearest(dropped, Work).Edge);
    }

    [Fact]
    public void The_middle_of_the_top_edge_is_a_half()
    {
        var middle = new Box((Work.Width - Width) / 2, 0, Width, Height);

        Assert.Equal(0.5, IslandPlacement.Nearest(middle, Work).Along, 3);
    }

    [Fact]
    public void The_place_is_kept_as_a_share_so_a_smaller_screen_keeps_it()
    {
        var position = new IslandPosition(ScreenEdge.Top, 0.5);

        var onBig = IslandPlacement.Place(position, Width, Height, Work);
        var onSmall = IslandPlacement.Place(position, Width, Height, new Box(0, 0, 1280, 700));

        Assert.Equal(Work.CentreX, onBig.CentreX, 3);
        Assert.Equal(640, onSmall.CentreX, 3);
    }

    [Fact]
    public void A_saved_place_comes_back_the_same()
    {
        var dropped = new Box(1500, 3, Width, Height);

        var position = IslandPlacement.Nearest(dropped, Work);
        var back = IslandPlacement.Place(position, Width, Height, Work);

        Assert.Equal(dropped.X, back.X, 3);
        Assert.Equal(Work.Y, back.Y, 3);
    }

    [Fact]
    public void The_island_is_pressed_against_its_edge()
    {
        Assert.Equal(Work.Y, IslandPlacement.Place(new IslandPosition(ScreenEdge.Top, 0.3), Width, Height, Work).Y, 3);
        Assert.Equal(Work.Bottom - Height, IslandPlacement.Place(new IslandPosition(ScreenEdge.Bottom, 0.3), Width, Height, Work).Bottom - Height, 3);
        Assert.Equal(Work.X, IslandPlacement.Place(new IslandPosition(ScreenEdge.Left, 0.3), Width, Height, Work).X, 3);
        Assert.Equal(Work.Right - Width, IslandPlacement.Place(new IslandPosition(ScreenEdge.Right, 0.3), Width, Height, Work).X, 3);
    }

    [Fact]
    public void An_island_wider_than_the_screen_has_nowhere_to_slide()
    {
        var narrow = new Box(0, 0, 100, 800);

        var placed = IslandPlacement.Place(new IslandPosition(ScreenEdge.Top, 0.8), Width, Height, narrow);

        Assert.Equal(narrow.X, placed.X, 3);
    }

    [Fact]
    public void A_share_outside_the_screen_is_pulled_back_in()
    {
        var placed = IslandPlacement.Place(new IslandPosition(ScreenEdge.Top, 4), Width, Height, Work);

        Assert.Equal(Work.Right - Width, placed.X, 3);
    }

    [Fact]
    public void Rings_stand_in_a_row_on_top_and_bottom_and_in_a_column_on_the_sides()
    {
        Assert.True(IslandPlacement.IsHorizontal(ScreenEdge.Top));
        Assert.True(IslandPlacement.IsHorizontal(ScreenEdge.Bottom));
        Assert.False(IslandPlacement.IsHorizontal(ScreenEdge.Left));
        Assert.False(IslandPlacement.IsHorizontal(ScreenEdge.Right));
    }

    [Fact]
    public void A_second_screen_to_the_left_has_negative_coordinates_and_still_works()
    {
        // Windows puts a monitor left of the main one at a negative x.
        var second = new Box(-1920, 0, 1920, 1080);
        var dropped = new Box(-1000, 8, Width, Height);

        var position = IslandPlacement.Nearest(dropped, second);
        var back = IslandPlacement.Place(position, Width, Height, second);

        Assert.Equal(ScreenEdge.Top, position.Edge);
        Assert.Equal(dropped.X, back.X, 3);
    }

    // ---- the island in hand ----

    [Theory]
    [InlineData(960, 30, ScreenEdge.Top)]
    [InlineData(960, 1000, ScreenEdge.Bottom)]
    [InlineData(40, 520, ScreenEdge.Left)]
    [InlineData(1880, 520, ScreenEdge.Right)]
    public void The_landing_strip_goes_to_the_edge_nearest_the_middle(double x, double y, ScreenEdge edge)
    {
        Assert.Equal(edge, IslandPlacement.NearestEdge(x, y, Work));
    }

    [Fact]
    public void Near_a_corner_the_edge_in_use_holds_until_another_is_clearly_closer()
    {
        // 60 from the top, 50 from the left: the left is closer, but by less than the stickiness.
        Assert.Equal(ScreenEdge.Top, IslandPlacement.NearestEdge(50, 60, Work, ScreenEdge.Top));
        Assert.Equal(ScreenEdge.Left, IslandPlacement.NearestEdge(50, 60, Work));

        // 80 from the top: now the left is closer by 30, and the strip moves.
        Assert.Equal(ScreenEdge.Left, IslandPlacement.NearestEdge(50, 80, Work, ScreenEdge.Top));
    }

    [Fact]
    public void Letting_go_keeps_the_place_along_the_edge_where_the_middle_was()
    {
        var position = IslandPlacement.DropAt(ScreenEdge.Top, 960, 200, Width, Height, Work);
        var placed = IslandPlacement.Place(position, Width, Height, Work);

        Assert.Equal(ScreenEdge.Top, position.Edge);
        Assert.Equal(960, placed.CentreX, 3);
        Assert.Equal(0, placed.Y);
    }

    [Fact]
    public void Letting_go_past_the_end_of_an_edge_stops_at_the_end()
    {
        var position = IslandPlacement.DropAt(ScreenEdge.Right, 1900, 1200, 34, 60, Work);

        Assert.Equal(1, position.Along);
        Assert.Equal(1040 - 60, IslandPlacement.Place(position, 34, 60, Work).Y, 3);
    }
}

File: mac/Tests/AikoKitTests/IslandPlacementTests.swift
import Testing

@testable import AikoKit

struct IslandPlacementTests {
    /// A 1920 by 1080 screen with the taskbar at the bottom.
    static let work = Box(0, 0, 1920, 1040)

    static let width = 160.0
    static let height = 34.0

    @Test
    func anIslandNearTheTopSticksToTheTop() {
        let dropped = Box(880, 12, Self.width, Self.height)

        #expect(IslandPlacement.nearest(dropped, Self.work).edge == .top)
    }

    @Test
    func anIslandNearTheRightSticksToTheRight() {
        let dropped = Box(1740, 500, Self.width, Self.height)

        #expect(IslandPlacement.nearest(dropped, Self.work).edge == .right)
    }

    @Test
    func anIslandNearTheBottomSticksToTheBottom() {
        let dropped = Box(400, 1000, Self.width, Self.height)

        #expect(IslandPlacement.nearest(dropped, Self.work).edge == .bottom)
    }

    @Test
    func theMiddleOfTheTopEdgeIsAHalf() {
        let middle = Box((Self.work.width - Self.width) / 2, 0, Self.width, Self.height)

        expectClose(0.5, IslandPlacement.nearest(middle, Self.work).along, places: 3)
    }

    @Test
    func thePlaceIsKeptAsAShareSoASmallerScreenKeepsIt() {
        let position = IslandPosition(.top, 0.5)

        let onBig = IslandPlacement.place(position, Self.width, Self.height, Self.work)
        let onSmall = IslandPlacement.place(position, Self.width, Self.height, Box(0, 0, 1280, 700))

        expectClose(Self.work.centreX, onBig.centreX, places: 3)
        expectClose(640, onSmall.centreX, places: 3)
    }

    @Test
    func aSavedPlaceComesBackTheSame() {
        let dropped = Box(1500, 3, Self.width, Self.height)

        let position = IslandPlacement.nearest(dropped, Self.work)
        let back = IslandPlacement.place(position, Self.width, Self.height, Self.work)

        expectClose(dropped.x, back.x, places: 3)
        expectClose(Self.work.y, back.y, places: 3)
    }

    @Test
    func theIslandIsPressedAgainstItsEdge() {
        expectClose(
            Self.work.y,
            IslandPlacement.place(IslandPosition(.top, 0.3), Self.width, Self.height, Self.work).y,
            places: 3)
        expectClose(
            Self.work.bottom - Self.height,
            IslandPlacement.place(IslandPosition(.bottom, 0.3), Self.width, Self.height, Self.work).bottom - Self.height,
            places: 3)
        expectClose(
            Self.work.x,
            IslandPlacement.place(IslandPosition(.left, 0.3), Self.width, Self.height, Self.work).x,
            places: 3)
        expectClose(
            Self.work.right - Self.width,
            IslandPlacement.place(IslandPosition(.right, 0.3), Self.width, Self.height, Self.work).x,
            places: 3)
    }

    @Test
    func anIslandWiderThanTheScreenHasNowhereToSlide() {
        let narrow = Box(0, 0, 100, 800)

        let placed = IslandPlacement.place(IslandPosition(.top, 0.8), Self.width, Self.height, narrow)

        expectClose(narrow.x, placed.x, places: 3)
    }

    @Test
    func aShareOutsideTheScreenIsPulledBackIn() {
        let placed = IslandPlacement.place(IslandPosition(.top, 4), Self.width, Self.height, Self.work)

        expectClose(Self.work.right - Self.width, placed.x, places: 3)
    }

    @Test
    func ringsStandInARowOnTopAndBottomAndInAColumnOnTheSides() {
        #expect(IslandPlacement.isHorizontal(.top))
        #expect(IslandPlacement.isHorizontal(.bottom))
        #expect(!IslandPlacement.isHorizontal(.left))
        #expect(!IslandPlacement.isHorizontal(.right))
    }

    @Test
    func aSecondScreenToTheLeftHasNegativeCoordinatesAndStillWorks() {
        // Windows puts a monitor left of the main one at a negative x, and so does macOS.
        let second = Box(-1920, 0, 1920, 1080)
        let dropped = Box(-1000, 8, Self.width, Self.height)

        let position = IslandPlacement.nearest(dropped, second)
        let back = IslandPlacement.place(position, Self.width, Self.height, second)

        #expect(position.edge == .top)
        expectClose(dropped.x, back.x, places: 3)
    }

    // ---- the island in hand ----

    @Test(arguments: [
        (960.0, 30.0, ScreenEdge.top),
        (960.0, 1000.0, ScreenEdge.bottom),
        (40.0, 520.0, ScreenEdge.left),
        (1880.0, 520.0, ScreenEdge.right),
    ])
    func theLandingStripGoesToTheEdgeNearestTheMiddle(x: Double, y: Double, edge: ScreenEdge) {
        #expect(IslandPlacement.nearestEdge(x, y, Self.work) == edge)
    }

    @Test
    func nearACornerTheEdgeInUseHoldsUntilAnotherIsClearlyCloser() {
        // 60 from the top, 50 from the left: the left is closer, but by less than the stickiness.
        #expect(IslandPlacement.nearestEdge(50, 60, Self.work, .top) == .top)
        #expect(IslandPlacement.nearestEdge(50, 60, Self.work) == .left)

        // 80 from the top: now the left is closer by 30, and the strip moves.
        #expect(IslandPlacement.nearestEdge(50, 80, Self.work, .top) == .left)
    }

    @Test
    func lettingGoKeepsThePlaceAlongTheEdgeWhereTheMiddleWas() {
        let position = IslandPlacement.dropAt(.top, 960, 200, Self.width, Self.height, Self.work)
        let placed = IslandPlacement.place(position, Self.width, Self.height, Self.work)

        #expect(position.edge == .top)
        expectClose(960, placed.centreX, places: 3)
        #expect(placed.y == 0)
    }

    @Test
    func lettingGoPastTheEndOfAnEdgeStopsAtTheEnd() {
        let position = IslandPlacement.dropAt(.right, 1900, 1200, 34, 60, Self.work)

        #expect(position.along == 1)
        expectClose(1040 - 60, IslandPlacement.place(position, 34, 60, Self.work).y, places: 3)
    }

    /// The landing strip reaches past the screen edge by its own corner radius, so the corners on
    /// that side fall outside the screen and the edge reads flat (D-186).
    @Test
    func theLandingStripReachesPastTheEdgeItSitsOn() {
        let strip = Box(100, 0, 200, 60)

        #expect(IslandPlacement.pastTheEdge(strip, .top, by: 8) == Box(100, -8, 200, 68))
        #expect(IslandPlacement.pastTheEdge(strip, .bottom, by: 8) == Box(100, 0, 200, 68))
        #expect(IslandPlacement.pastTheEdge(strip, .left, by: 8) == Box(92, 0, 208, 60))
        #expect(IslandPlacement.pastTheEdge(strip, .right, by: 8) == Box(100, 0, 208, 60))
    }
}
